compiler/rustc_data_structures/src/thin_vec.rs

```diff
@@ -46,6 +56,102 @@ impl<T> ::std::ops::DerefMut for ThinVec<T> {
     }
 }
 
+impl<T> FromIterator<T> for ThinVec<T> {
+    fn from_iter<I: IntoIterator<Item = T>>(iter: I) -> Self {
+        let mut iter = iter.into_iter();
+
+        match iter.next() {
+            None => ThinVec(None),
+            Some(first) => {
+                // + 1 for the first element
+                let mut vec = Vec::with_capacity(iter.size_hint().0 + 1);
+
+                vec.push(first);
+                for next in iter {
+                    vec.push(next);
+                }
+
+                ThinVec(Some(Box::new(vec)))
+            }
+        }
+    }
+}
+
+impl<T> IntoIterator for ThinVec<T> {
+    type Item = T;
+    type IntoIter = IntoIterOwned<T>;
+
+    fn into_iter(self) -> IntoIterOwned<T> {
+        match self {
+            ThinVec(None) => IntoIterOwned(None),
+            ThinVec(Some(vec)) => IntoIterOwned(Some(vec.into_iter())),
+        }
```

Comment:
You could avoid creating custom iterators by returning
```rust
self.unwrap_or_else(Vec::new).into_iter()
```

Reply:
Yes, but that would require an allocation in the empty case, which defeats the primary purpose of these impls.

Reply:
Vec::new does not allocate.

Reply:
Although, we could consider using `self.unwrap_or_else(|| Vec::with_capacity(0)).into_iter()`, since the docs for `with_capacity` say that will avoid allocating. However, it might be better to just keep the custom iterators since they may still be faster.

Reply:
> Vec::new does not allocate.

Huh, I did not realize that! So, does `ThinVec` only exist to save 16 bytes of space in the types that it is a part of?

Reply:
Is there any chance that the extra space taken up by using `Vec::new` is significant enough that it would be slower to use your version, or no?

Reply:
> Huh, I did not realize that! So, does ThinVec only exist to save 16 bytes of space in the types that it is a part of?

Yes.

> Is there any chance that the extra space taken up by using `Vec::new` is significant enough that it would be slower to use your version, or no?

I don't know. I expect that my version would be slightly faster and simpler. Wrapping the iterator in an Option adds a boolean and a branch and a test to each call to `next()`, without saving any space.

Reply:
Okay, thanks for the suggestion! I'll use your approach.

Reply:
Hmm, I'm getting borrowck errors about returning a reference to a local variable. (It works for the owned iterator, but not for the by-reference ones.) Is there some way to solve this that I'm not aware of, or do I need the custom iterators after all?

I'm wondering if this is the reason I used the custom iterators in the first place...

Reply:
For the by-ref iterator, you can use something like:
```rust
self.as_deref().unwrap_or(&[]).iter()
```

Reply:
Actually, just using

```rust
self.as_ref().iter()
```

seems to work.